File: JavaScript/src/Internals/Functions/Csharp/Function_TRIMEND.js
import { Function_2 } from '../Function_2.js';
import { Operand } from '../../../Operand.js';
import { StringCache } from '../../../Internals/StringCache.js';

/**
 * Function_TRIMEND
 */
export class Function_TRIMEND extends Function_2 {

    get Name() {
        return "TrimEnd";
    }
    
    constructor(z) {
    super(z);
  }
    
    /**
     * @param {AlgorithmEngine} engine
     * @returns {Operand}
     */
    Evaluate(engine, tempParameter) {
        let args1 = this.GetText_1(engine, tempParameter);
        if (args1.IsError) { return args1; }

        if (this.b == null) {
            return Operand.Create(args1.TextValue.trimEnd());
        }
        
        let args2 = this.GetText_2(engine, tempParameter);
        if (args2.IsError) { return args2; }
        
        let trimChars = args2.TextValue.split('');
        let text = args1.TextValue;
        let index = text.length - 1;
        
        while (index >= 0 && trimChars.includes(text[index])) {
            index--;
        }
        
        return Operand.Create(text.substring(0, index + 1));
    }
    

}

File: java/toolgood.algorithm/src/main/java/toolgood/algorithm/internals/Functions/csharp/Function_TRIMEND.java
package toolgood.algorithm.internals.functions.csharp;

import toolgood.algorithm.Operand;
import toolgood.algorithm.internals.AlgorithmEngine;
import toolgood.algorithm.internals.functions.Function_2;
import toolgood.algorithm.internals.functions.FunctionBase;

public class Function_TRIMEND extends Function_2 {
    public Function_TRIMEND(FunctionBase func1, FunctionBase func2) {
        super(func1, func2);
    }

    @Override
    public Operand evaluate(AlgorithmEngine work, java.util.function.Function<String, Operand> tempParameter) {
        Operand args1 = func1.evaluate(work, tempParameter); if(args1.isNotText()) { args1 = args1.toText("Function '{0}' parameter {1} is error!", "TrimEnd", 1); if(args1.isError()) return args1; }
        if(func2 == null) {
            return Operand.create(args1.getTextValue().trim());
        }
        Operand args2 = func2.evaluate(work, tempParameter); if(args2.isNotText()) { args2 = args2.toText("Function '{0}' parameter {1} is error!", "TrimEnd", 2); if(args2.isError()) return args2; }
        char[] trimChars = args2.getTextValue().toCharArray();
        return Operand.create(trimEnd(args1.getTextValue(), trimChars));
    }

    private String trimEnd(String str, char[] trimChars) {
        if (str == null || str.isEmpty()) {
            return str;
        }
        int length = str.length();
        while (length > 0 && contains(trimChars, str.charAt(length - 1))) {
            length--;
        }
        return str.substring(0, length);
    }

    private boolean contains(char[] array, char c) {
        for (char ch : array) {
            if (ch == c) {
                return true;
            }
        }
        return false;
    }

    @Override
    public void toString(StringBuilder stringBuilder, boolean addBrackets) {
        addFunction(stringBuilder, "TrimEnd");
    }
}
